### PR TITLE
bpo-14935: Remove static state from the _csv module

### PR DESCRIPTION
**Pull request by encukou**
This merges code/ideas from gh-20974 and gh-16078, and uses the new `_PyType_GetModuleByDef`.

Co-authored-by: Marcel Plch <gmarcel.plch@gmail.com>
Co-authored-by: Eddie Elizondo <eduardo.elizondorueda@gmail.com>
Co-authored-by: Hai Shi <shihai1992@gmail.com>

https://bugs.python.org/issue14935

### CODE REVIEW

File: Modules/_csv.c
Comment (shihai1991) [inline Modules/_csv.c]: nitpick: Victor added a new perfect  `PyModule_AddObjectRef()` function in https://github.com/python/cpython/pull/23122
MAYBE you can try it in your PR :)

File: Modules/_csv.c
Comment (shihai1991) [inline Modules/_csv.c:1519]: This slots can be removed if there have no actual slot? something like `PyModuleDef`. For example: https://github.com/python/cpython/blob/master/Modules/_bisectmodule.c#L286-L292
Pter, Do you think it worth to be enhanced? MAYBE I can create a new bpo to enhance it :)

Comment (encukou) [reply Modules/_csv.c:1519]: It would need a new BPO, yes. But I don't recommend it: if people start omitting slots in the stable ABI, their extensions will fail on older Python versions :(

File: Modules/_csv.c
Comment (encukou) [reply Modules/_csv.c:1670]: I believe the tests wouldn't pass, because ` PyModule_AddObjectRef` is not part of the stable ABI on Windows (`PC/python3dll.c`).

Comment (encukou) [reply Modules/_csv.c]: Let me try this!

File: Modules/_csv.c
Comment (shihai1991) [reply Modules/_csv.c:1519]: > But I don't recommend it: if people start omitting slots in the stable ABI, their extensions will fail on older Python versions :(

Oh, you are right. it's a problem~

File: Modules/_csv.c
Comment (vstinner) [inline Modules/_csv.c]: I suggest to name the variable just "state". m_state is the name of the PyModuleObject member. I understood that Python structure member names have a prefix to help to guess the type of a variable. But I don't think that it's useful here.

File: Modules/_csv.c
Comment (vstinner) [inline Modules/_csv.c]: I may be safer to use Py_CLEAR() here, no? A finalized object *can* be ressurected. Same remark for other finalizer functions below.

File: Modules/_csv.c
Comment (vstinner) [inline Modules/_csv.c]: nitpick: I prefer longer variable names like "mod" or "module", rather than just "m" :-)

File: Modules/_csv.c
Comment (vstinner) [inline Modules/_csv.c]: You may write a subfunction to do that in a single function, to factorize the code (since you have similar code below). It would raise the exception as well. So here, you would only have to "return NULL;" on failure.

If you want to keep "dialect_new" in the error message, you can pass it as an argument.

File: Modules/_csv.c
Comment (vstinner) [inline Modules/_csv.c:906]: I suggest to add `self->field = NULL;` to reduce the risk of crash if the object is resurected.

File: Modules/_csv.c
Comment (vstinner) [inline Modules/_csv.c]: It seems like the caller usually already gets the state, so maybe pass the state directly rather than the module?

File: Modules/_csv.c
Comment (vstinner) [inline Modules/_csv.c:1471]: I have a personal preference to only read the state once, and so add a "state" variable.

File: Modules/_csv.c
Comment (vstinner) [reply Modules/_csv.c:1519]: > It would need a new BPO, yes. But I don't recommend it: if people start omitting slots in the stable ABI, their extensions will fail on older Python versions :(

But the _csv doesn't use the limited C API right now. Well, it's up to you ;-)

File: Modules/_csv.c
Comment (encukou) [reply Modules/_csv.c]: It stands for “module state”. I'll rename it to `module_state` to make it clearer.

File: Modules/_csv.c
Comment (encukou) [reply Modules/_csv.c:1519]: _csv doesn't use it, but if `PyType_FromSpec*` started accepting `.slots = NULL`, modules that *do* use the limited API could get the problem.
I'm thinking about a solution, but that's for a PEP, not this PR :)

File: Modules/_csv.c
Comment (vstinner) [inline Modules/_csv.c]: nitpick, but you can leave the code as it it.

```suggestion
    self->error_obj = Py_NewRef(module_state->error_obj);    
```

File: Modules/_csv.c
Comment (vstinner) [inline Modules/_csv.c:353]: I would expect that _csv_state_from_type() cannot fail (return NULL) and so use assertions rather than runtime checks (if).

If you consider that it *can* fail, you should check if _csv_state_from_type() fails at each _csv_state_from_type() call, no? For example, dialect_new() uses module_state without checking if it is NULL.

I didn't check if _csv_state_from_type() can be called on a type which is a subtype of the *type* parameter.

Well, in case of doubt, just check for _csv_state_from_type() failure ;-)

File: Modules/_csv.c
Comment (vstinner) [inline Modules/_csv.c:415]: Technically, Py_XDECREF is just fine, there is no need to set dialect to NULL, since the dialect local scope if this function, and we exit the scope (function) at the next line. Same remark for the "err" label. But it's up to you, it's just a remark ;-)

File: Modules/_csv.c
Comment (encukou) [reply Modules/_csv.c:353]: > I didn't check if _csv_state_from_type() can be called on a type which is a subtype of the type parameter.

It can; that's why this uses `_PyType_GetModuleByDef`.
I'll add the missing NULL check.

File: Modules/_csv.c
Comment (encukou) [reply Modules/_csv.c:415]: I'll just leave that up to the compiler :)

File: Modules/_csv.c
Comment (vstinner) [inline Modules/_csv.c]: maybe use ".m_xxx = xxx" syntax here, as you did for the other spec below. Same remark for Reader_Type_spec.

**Review by vstinner — APPROVED** [review]
LGTM. IMO the code is cleaner with your changes, thanks.

I just added a minor coding style remark, feel free to ignore it ;-)

**Review by vstinner — APPROVED** [review]
LGTM, thanks.

**Review by shihai1991 — APPROVED** [review]
I have no other comments, so LGTM.

**Pull request merged**
